util/messaging.py

```diff
@@ -0,0 +1,234 @@
+        # inbound queue for receiving messages (inbound calls and responses of outbound calls)
+        inbound_queue = tf.FIFOQueue(queue_size, [tf.string], name=self.id, shared_name=self.id)
+        self._inbound_dequeue = inbound_queue.dequeue()
+        self._inbound_close = inbound_queue.close(cancel_pending_enqueues=True)
+        # outbound queues for sending messages (outbound calls and responses to inbound calls)
+        self._outbound_enqueues = {}
+        self._outbound_closes = []
+        for j in cluster_spec.jobs:
+            for t in cluster_spec.task_indices(j):
+                id = self._id_pattern % (j, t)
+                # skip local inbound queue
+                if id != self.id:
+                    queue = tf.FIFOQueue(queue_size, [tf.string], name=id, shared_name=id)
+                    self._outbound_enqueues[id] = queue.enqueue([self._ph_content])
+                    self._outbound_closes.append(queue.close(cancel_pending_enqueues=True))
+        # response lookup for outbound RPC calls (that are waiting for a response)
+        self._waiting_for_response = {}
+        # response message id counter and lock
+        self._response_id_counter = 0
+        self._response_id_lock = Lock()
+        # local-only outbound queue for decoupling (otherwise blocking) send/receive dependencies
+        self._to_send = Queue.Queue()
+
+    def _call(self, function, args):
+        '''
+        Internal function to call a local class method as RPC function.
+        'function' - Name of the local function (member of this class)
+        'args' - Tuple of call arguments
+        '''
+        try:
+            fun = getattr(self, function, *args)
+        except AttributeError:
+            log.warn('Function not found: %s, arguments: %r' % (function, args))
+            return
+        return fun(*args)
+
+    def _send_loop(self, session, coord):
+        '''
+        Performs outbound messaging loop.
+        'session' - The current session
+        'coord' - A thread coordinator to automatically stop the loop
+        '''
+        while not coord.should_stop():
+            # get enqueue and feed_dict from local sending queue
+            enqueue, feed_dict = self._to_send.get()
+            try:
+                content = session.run(enqueue, feed_dict=feed_dict)
+            except (tf.errors.CancelledError, tf.errors.OutOfRangeError):
+                # queue got closed
+                return
+
+    def _receive(self, session, caller, is_response, response_id, function, items):
+        '''
+        Executes an inbound message. Either by calling a local RPC member function or by
+        continuing a halted/callback-ed local call that is waiting for a response.
+        'session' - The current session
+        'caller' - Id of the caller (used to respond in case of inbound RPC call)
+        'is_response' - If the message is a response (and not an inbound RPC call)
```

Comment:
It feels to me that `is_response` is superflous, and could be just `is_response = caller >= 0` ?

Reply:
Caller is related to the current message sender - therefore it represents different roles during a RPC session. It should also be kept to be able to log accordingly.